source/gateway.go

```diff
@@ -0,0 +1,549 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"text/template"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeinformers "k8s.io/client-go/informers"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	cache "k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gateway "sigs.k8s.io/gateway-api/pkg/client/clientset/gateway/versioned"
+	informers "sigs.k8s.io/gateway-api/pkg/client/informers/gateway/externalversions"
+	informers_v1a2 "sigs.k8s.io/gateway-api/pkg/client/informers/gateway/externalversions/apis/v1alpha2"
+
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+const (
+	gatewayGroup = "gateway.networking.k8s.io"
+	gatewayKind  = "Gateway"
+)
+
+type gatewayRoute interface {
+	// Object returns the underlying route object to be used by templates.
+	Object() kubeObject
+	// Metadata returns the route's metadata.
+	Metadata() *metav1.ObjectMeta
+	// Hostnames returns the route's specified hostnames.
+	Hostnames() []v1alpha2.Hostname
+	// Protocol returns the route's protocol type.
+	Protocol() v1alpha2.ProtocolType
+	// RouteStatus returns the route's common status.
+	RouteStatus() v1alpha2.RouteStatus
+}
+
+type newGatewayRouteInformerFunc func(informers.SharedInformerFactory) gatewayRouteInformer
+
+type gatewayRouteInformer interface {
+	List(namespace string, selector labels.Selector) ([]gatewayRoute, error)
+	Informer() cache.SharedIndexInformer
+}
+
+func newGatewayInformerFactory(client gateway.Interface, namespace string, labelSelector labels.Selector) informers.SharedInformerFactory {
+	var opts []informers.SharedInformerOption
+	if namespace != "" {
+		opts = append(opts, informers.WithNamespace(namespace))
+	}
+	if labelSelector != nil && !labelSelector.Empty() {
+		lbls := labelSelector.String()
+		opts = append(opts, informers.WithTweakListOptions(func(o *metav1.ListOptions) {
+			o.LabelSelector = lbls
+		}))
+	}
+	return informers.NewSharedInformerFactoryWithOptions(client, 0, opts...)
+}
+
+type gatewayRouteSource struct {
+	gwNamespace string
+	gwLabels    labels.Selector
+	gwInformer  informers_v1a2.GatewayInformer
+
+	rtKind        string
+	rtNamespace   string
+	rtLabels      labels.Selector
+	rtAnnotations labels.Selector
+	rtInformer    gatewayRouteInformer
+
+	nsInformer coreinformers.NamespaceInformer
+
+	fqdnTemplate             *template.Template
+	combineFQDNAnnotation    bool
+	ignoreHostnameAnnotation bool
+}
+
+func newGatewayRouteSource(clients ClientGenerator, config *Config, kind string, newInformerFn newGatewayRouteInformerFunc) (Source, error) {
+	ctx := context.TODO()
```

Comment:
Is `context.TODO` the right context to use here?

Reply:
I think so.

https://pkg.go.dev/context#TODO

> TODO returns a non-nil, empty Context. Code should use context.TODO when it's unclear which Context to use or it is not yet available (because the surrounding function has not yet been extended to accept a Context parameter).